### DISCUSSION TITLE
What if I have SObjects.cls implemented already?

### DISCUSSION THREAD

**Discussion by PeterLin888** [wimvelzeboer/fflib-apex-extensions#6] [Q&A] [answered]
Hi William,

I do not have SObjects.cls implemented.

Let's suppose that I am working on a different org and I do have SObjects.cls implemented. And I have thousands of lines of codes depending on this SObjects.cls.

As your "How to adapt" link below, I would need to follow your instruction to extend all my Domain classes to SObjects.
https://github.com/wimvelzeboer/fflib-apex-extensions/blob/main/docs/how-to-adapt.md

And let's suppose I really like to package this repository as an SFDX Unlocked Package (SFDXUP) version to install.

Which option would you suggest that I should implement this repository "fflib-apex-extensions"?
1. Refactor thousands of lines of codes depending on this SObjects.cls to have them all depend on a different name, e.g. SObjectsOLD.cls. Then every time when there are more feature or bug-fix changes in SObjects.cls in this repository, I will just package a new SFDXUP version and then install it.
2. Do not refactor my SObjects.cls. Instead, refactor SObjects.cls in this repository as a different name, e.g. SObjectsNEW.cls. Then every time when there are more feature or bug-fix changes in SObjects.cls in this repository, anyone in my team and I will have to **remember** to refactor SObjects.cls as SObjectsNEW.cls before packaging a new SFDXUP version and then install it.
3. Will you consider refactor SObjects.cls to fflib2_SObjects.cls or something similar, e.g. fflib_SObjects2.cls, so that I (or someone else) don't have to worry about 1 or 2 above?
4. Let's suppose that I don't have the courage to do 1 or 2 above. So, I will give up and do not dare to think about implementing this repository anymore.
5. Let's suppose that someday when more and more teams have issues of 1 or 2 above and you choose to add or change a different prefix. Then all teams who have installed their the SFDXUP versions have to refactor their SFDXUP versions because of the different prefix to be implemented.

I guess that probably this question may apply to some other Classes in this repository as below. (Probably just adding fflib_ prefix.)
Dates.cls
DateTimes.cls
Decimals.cls
Doubles.cls
Ids.cls
Integers.cls
Longs.cls
Objects.cls
PrimitiveDomainsTest.cls
Strings.cls

I will be more comfortable to package this as a SFDXUP version if I can be sure that nobody needs to worry about doing 1 or 2 or 5 above either now or in the future.

Thanks for this excellent fflib-apex-extensions repository.

**Comment by wimvelzeboer** [comment, marked as answer]
The more I think about this the more I agree with you. 
There are also a number of PR open for a very long time in the [fflib-apex-common](https://github.com/apex-enterprise-patterns/fflib-apex-common/pulls) repo, a bit frustrating as it is blocking my progress.

So, at this moment I am thinking it is indeed a good idea of adding "fflib_" as prefix. And moving some code (of those open PRs) to files with a name like: `fflib_Objects2`, `fflib_SObject2`, `fflib_SObjectSelector2`, and do something similar in the future for code that I would like to move to fflib-apex-common but still in PR awaiting approval.

Then we would get something like
```
 public virtual class fflib_Object2 extends fflib_Object
 {
    /// code awaiting approval
}
```
This file would be empty when all code is merged into fflib-apex-common, and would then be a place holder for new additions.
Would that be an idea?

**Reply by PeterLin888** [reply]
Thank you so much, William!

The changes you mentioned have solved all my concerns and worries in my original comment above.

I think I am getting ready to package this repository as my next official internal Unlocked Package.

I would be happy to help on adding more code coverage and do some refactoring (starting from adding fflib_).

**Comment by wimvelzeboer** [comment]
@PeterLinBiz I have created PR #14 with a change to add the "fflib_" prefix. Was this your main concern or is there now still anything else?

**Reply by PeterLin888** [reply]
@wimvelzeboer, Thank you so much! The PR #14 is great for adding the "fflib_". I added one comment at PR #14 .
I have no more other concern about this repository. Probably for the next steps, I will try to add some more code coverage.
Probably you may consider making a prioritized list so that we (assuming more people will come to help) may start to work on the items on the top of the list.
